File: Android/app/src/main/java/com/auralis/crisisconnect/messaging/ResourceAlertWakeClient.kt
package com.auralis.crisisconnect.messaging

import android.content.Context
import android.content.Intent
import com.auralis.crisisconnect.BuildConfig
import com.auralis.crisisconnect.network.PinnedOkHttpClient
import com.google.firebase.auth.FirebaseAuth
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

internal data class ResourceAlertWakePayload(
    val panelId: String,
    val attemptId: String,
    val receiptNonce: String,
)

private val RESOURCE_PANEL_ID = Regex("^[a-z0-9._-]{1,96}$")
private val RESOURCE_ATTEMPT_ID = Regex("^[A-Za-z0-9:_-]{1,128}$")
private val RESOURCE_RECEIPT_NONCE = Regex("^[A-Za-z0-9_-]{16,128}$")

internal fun parseResourceAlertWake(data: Map<String, String>): ResourceAlertWakePayload? {
    if (data["type"] != "resource_alert_wake") return null
    val panelId = data["panelId"]?.trim().orEmpty()
    val attemptId = data["attemptId"]?.trim().orEmpty()
    val receiptNonce = data["receiptNonce"]?.trim().orEmpty()
    if (!RESOURCE_PANEL_ID.matches(panelId) || !RESOURCE_ATTEMPT_ID.matches(attemptId) ||
        !RESOURCE_RECEIPT_NONCE.matches(receiptNonce)
    ) return null
    return ResourceAlertWakePayload(panelId, attemptId, receiptNonce)
}

internal fun buildResourceAlertWakeAck(payload: ResourceAlertWakePayload): JSONObject = JSONObject()
    .put("panelId", payload.panelId)
    .put("action", "ackWake")
    .put("attemptId", payload.attemptId)
    .put("receiptNonce", payload.receiptNonce)
    .put("source", "native")

internal object ResourceAlertWakeClient {
    private const val DEFAULT_BASE_URL = "https://crisisconnect.network"
    private const val REFRESH_ACTION = "com.auralis.crisisconnect.RESOURCE_ALERT_INBOX_REFRESH"
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    suspend fun acknowledge(context: Context, data: Map<String, String>): Boolean {
        val payload = parseResourceAlertWake(data) ?: return false
        val user = FirebaseAuth.getInstance().currentUser?.takeUnless { it.isAnonymous } ?: return false
        val queue = ResourceAlertWakeQueue.shared(context)
        val now = System.currentTimeMillis()
        val key = queue.enqueue(payload, user.uid, now) ?: return false
        ResourceAlertWakeAckWorker.enqueue(context)
        val claimed = queue.claim(user.uid, now, key) ?: return false
        val acknowledged = deliver(user.uid, context, claimed)
        if (queue.hasPending(user.uid, System.currentTimeMillis())) {
            ResourceAlertWakeAckWorker.enqueue(context)
        }
        return acknowledged
    }

    suspend fun drainPending(context: Context, maximumItems: Int = 8): Boolean {
        val user = FirebaseAuth.getInstance().currentUser?.takeUnless { it.isAnonymous } ?: return false
        val queue = ResourceAlertWakeQueue.shared(context)
        var processed = 0
        while (processed < maximumItems.coerceIn(1, 32)) {
            val claimed = queue.claim(user.uid, System.currentTimeMillis()) ?: break
            deliver(user.uid, context, claimed)
            processed += 1
        }
        return !queue.hasPending(user.uid, System.currentTimeMillis())
    }

    private suspend fun deliver(
        recipientUid: String,
        context: Context,
        pending: PendingResourceAlertWake,
    ): Boolean {
        val user = FirebaseAuth.getInstance().currentUser
            ?.takeUnless { it.isAnonymous }
            ?.takeIf { it.uid == recipientUid }
            ?: return false
        val endpoint = endpointUrl() ?: return false
        val token = runCatching { user.getIdToken(false).await().token?.trim() }.getOrNull()
            ?.takeIf { it.isNotBlank() }
            ?: return false
        var status = post(endpoint, token, pending.payload)
        if (status == 401) {
            val refreshed = runCatching { user.getIdToken(true).await().token?.trim() }.getOrNull()
                ?.takeIf { it.isNotBlank() }
                ?: return false
            status = post(endpoint, refreshed, pending.payload)
        }
        if (status in 200..299) {
            if (!ResourceAlertWakeQueue.shared(context).complete(
                    recipientUid,
                    pending.key,
                    System.currentTimeMillis(),
                )
            ) return false
            context.applicationContext.sendBroadcast(
                Intent(REFRESH_ACTION)
                    .setPackage(context.packageName)
                    .putExtra("panelId", pending.panelId)
            )
            return true
        }
        if (isPermanentFailure(status)) {
            ResourceAlertWakeQueue.shared(context).complete(
                recipientUid,
                pending.key,
                System.currentTimeMillis(),
            )
        }
        return false
    }

    private fun isPermanentFailure(status: Int): Boolean = status in 400..499 &&
        status !in setOf(401, 408, 425, 429)

    private suspend fun post(endpoint: HttpUrl, token: String, payload: ResourceAlertWakePayload): Int =
        withContext(Dispatchers.IO) {
            val body = buildResourceAlertWakeAck(payload).toString().toRequestBody(jsonMediaType)
            val request = Request.Builder()
                .url(endpoint)
                .header("Authorization", "Bearer $token")
                .header("Content-Type", "application/json")
                .post(body)
                .build()
            runCatching {
                PinnedOkHttpClient.newClient().newCall(request).execute().use { it.code }
            }.getOrDefault(503)
        }

    private fun endpointUrl(): HttpUrl? {
        val base = BuildConfig.MOBILE_SYNC_BASE_URL.trim().trimEnd('/').ifBlank { DEFAULT_BASE_URL }
        val parsed = base.toHttpUrlOrNull() ?: return null
        if (!parsed.isHttps && !(BuildConfig.DEBUG && isLocalDebugHost(parsed.host))) return null
        return parsed.newBuilder()
            .addPathSegments("api/dashboard/resource-alert-inbox")
            .build()
    }

    private fun isLocalDebugHost(host: String): Boolean = host == "10.0.2.2" ||
        host == "127.0.0.1" || host == "localhost" || host.endsWith(".local")
}

File: iOS/Crisis Connect/Services/ResourceAlerts/ResourceAlertWakeClient.swift
import FirebaseAuth
import Foundation

struct ResourceAlertWakePayload: Equatable, Sendable {
    let panelId: String
    let attemptId: String
    let receiptNonce: String

    init(panelId: String, attemptId: String, receiptNonce: String) {
        self.panelId = panelId
        self.attemptId = attemptId
        self.receiptNonce = receiptNonce
    }

    init?(userInfo: [AnyHashable: Any]) {
        func string(_ key: String) -> String? {
            (userInfo[key] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        guard string("type") == "resource_alert_wake",
              let panelId = string("panelId"),
              let attemptId = string("attemptId"),
              let receiptNonce = string("receiptNonce"),
              Self.matches(panelId, pattern: "[a-z0-9._-]{1,96}"),
              Self.matches(attemptId, pattern: "[A-Za-z0-9:_-]{1,128}"),
              Self.matches(receiptNonce, pattern: "[A-Za-z0-9_-]{16,128}") else {
            return nil
        }
        self.panelId = panelId
        self.attemptId = attemptId
        self.receiptNonce = receiptNonce
    }

    func makeRequest(endpoint: URL, idToken: String) throws -> URLRequest {
        let body: [String: String] = [
            "panelId": panelId,
            "action": "ackWake",
            "attemptId": attemptId,
            "receiptNonce": receiptNonce,
            "source": "native",
        ]
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(idToken)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        request.timeoutInterval = 10
        return request
    }

    private static func matches(_ value: String, pattern: String) -> Bool {
        value.range(of: "^(?:\(pattern))$", options: .regularExpression) != nil
    }
}

enum ResourceAlertWakeHandlingResult: Equatable, Sendable {
    case notResourceWake
    case acknowledged
    case queued
    case failed
}

extension Notification.Name {
    static let resourceAlertInboxRefreshRequested = Notification.Name(
        "crisisconnect.resourceAlertInboxRefreshRequested"
    )
}

final class ResourceAlertWakeClient: @unchecked Sendable {
    static let shared = ResourceAlertWakeClient()

    private let session: URLSession
    private let queue: ResourceAlertWakeQueue

    init(session: URLSession = .shared, queue: ResourceAlertWakeQueue = .shared) {
        self.session = session
        self.queue = queue
    }

    func handleRemotePush(userInfo: [AnyHashable: Any]) async -> ResourceAlertWakeHandlingResult {
        guard let payload = ResourceAlertWakePayload(userInfo: userInfo) else {
            return .notResourceWake
        }
        guard let user = Auth.auth().currentUser,
              !user.isAnonymous else {
            return .failed
        }
        do {
            let key = try await queue.enqueue(payload: payload, recipientUid: user.uid)
            let result = await deliverQueued(key: key, user: user)
            ResourceAlertWakeBackgroundManager.shared.scheduleIfNeeded()
            return result ? .acknowledged : .queued
        } catch {
            NSLog("ResourceAlertWakeClient: could not persist ACK")
            return .failed
        }
    }

    @discardableResult
    func drainPending(maximumItems: Int = 8) async -> Bool {
        guard let user = Auth.auth().currentUser, !user.isAnonymous else { return false }
        for _ in 0..<max(1, min(maximumItems, 32)) {
            let pending: PendingResourceAlertWake?
            do {
                pending = try await queue.claim(recipientUid: user.uid)
            } catch {
                NSLog("ResourceAlertWakeClient: could not read pending ACK queue")
                return false
            }
            guard let pending else { break }
            _ = await deliver(claimed: pending, user: user)
        }
        return (try? await queue.hasPending(recipientUid: user.uid)) == false
    }

    func nextPendingAttemptDate() async -> Date? {
        guard let user = Auth.auth().currentUser, !user.isAnonymous else { return nil }
        return try? await queue.nextAttemptDate(recipientUid: user.uid)
    }

    private func deliverQueued(key: String, user: User) async -> Bool {
        guard let claimed = try? await queue.claim(key: key, recipientUid: user.uid) else {
            return false
        }
        return await deliver(claimed: claimed, user: user)
    }

    private func deliver(claimed: PendingResourceAlertWake, user: User) async -> Bool {
        guard let endpoint = Self.endpointURL() else { return false }
        let status: Int
        do {
            let token = try await user.getIDToken()
            var response = try await send(payload: claimed.payload, endpoint: endpoint, idToken: token)
            if response == 401 {
                let refreshed = try await user.getIDToken(forcingRefresh: true)
                response = try await send(payload: claimed.payload, endpoint: endpoint, idToken: refreshed)
            }
            status = response
        } catch {
            NSLog("ResourceAlertWakeClient: ACK transport failed")
            return false
        }
        if (200...299).contains(status) {
            try? await queue.complete(key: claimed.key, recipientUid: user.uid)
            NotificationCenter.default.post(
                name: .resourceAlertInboxRefreshRequested,
                object: nil,
                userInfo: ["panelId": claimed.panelId]
            )
            return true
        }
        if Self.isPermanentFailure(status) {
            // The server has definitively rejected or retired this challenge; retrying it would only
            // leak battery and rate-limit budget. The durable personal inbox remains authoritative.
            try? await queue.complete(key: claimed.key, recipientUid: user.uid)
        }
        NSLog("ResourceAlertWakeClient: ACK rejected status=%d", status)
        return false
    }

    private static func isPermanentFailure(_ status: Int) -> Bool {
        (400...499).contains(status) && ![401, 408, 425, 429].contains(status)
    }

    private func send(payload: ResourceAlertWakePayload, endpoint: URL, idToken: String) async throws -> Int {
        let request = try payload.makeRequest(endpoint: endpoint, idToken: idToken)
        let (_, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { return 503 }
        return http.statusCode
    }

    private static func endpointURL() -> URL? {
        let raw = (Bundle.main.object(forInfoDictionaryKey: "CRISIS_CONNECT_WEB_URL") as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let base = raw.isEmpty ? "https://crisisconnect.network" : raw
        guard var components = URLComponents(string: base), components.scheme == "https" else {
            return nil
        }
        components.path = components.path.hasSuffix("/")
            ? components.path + "api/dashboard/resource-alert-inbox"
            : components.path + "/api/dashboard/resource-alert-inbox"
        components.query = nil
        components.fragment = nil
        return components.url
    }
}
